### ISSUE TITLE
#ITEM 2 - Ristretto Implementation over Sean's Doppio Curve

### ISSUE THREAD

**Issue by CPerezz** [dusk-network/dusk-zerocaf#76]
Since the curve that Sean provided in:

<blockquote class="twitter-tweet"><p lang="en" dir="ltr">Here&#39;s an &quot;embedded&quot; curve over ristretto255&#39;s scalar field<br><br>-x^2 + y^2 = 1 - (86649/86650)x^2y^2<br><br>which is Ristretto-ready and birationally equivalent to<br><br>y^2 = x^3 + 346598x^2 + x (and it&#39;s twist secure)<br><br>Any other suggestions?</p>&mdash; Sean Bowe (@ebfull) <a href="https://twitter.com/ebfull/status/1087571257057406976?ref_src=twsrc%5Etfw">January 22, 2019</a></blockquote> 

Provides values for the Twisted Edwards form that aren't suitable for a Ristretto protocol implementation we have two options:

- Find an isomorphic twist that allows us to mantain the same orders for the Finite Field and also the Sub group, and at the same time, gives `a & d` values that are suitable  for a Ristretto implementation.

- Choice a diferent curve. This will force us to:
    - Find the new order of the Sub group and refactor the Scalar implementation to work over the new order and implement Decaf or Ristretto depending on the co-factor that the new curve provides.

It seems that @Bounce23 found an isomorphic twist that can do the job. We will update here the discussions.

**Comment by CPerezz** [comment]
@Bounce23 provided the values: `a = -1`, `d = -86649` related to the isomorphic twist solution.

**Comment by CPerezz** [comment]
As far as we've seen, whith the implementation of `sqrt_ratio_i` and the refactor of `inv_sqrt` functions done on 9c00824 and 73ee720 and the refactor done then of compression/decompression functions done in 2f395a71, allowed us to pass:
- `bad_encoding_test`, see: https://ristretto.group/test_vectors/ristretto255.html
- `identity_point_encoding/decoding`.

We are still unable to pass the `good_encoding_tests` since the `RISTRETTO_BASEPOINT_COMPRESSED` variable cannot be decompressed (which might mean that we need to find our own Basepoint).

**Comment by CPerezz** [comment]
@Bounce23 found the following parameters for an isomorphic twist that should work:
```
sage: d = 86649

sage: a = -1

sage: b = d/a

sage: b = 4*b

sage: c = 2-b

sage: c

346598
```

This ones, will provide the same Sub-group order, and also satisfy:
- Twisted Edwards Curve with Co-factor 8.
- Satisfy the isogeny proof `b` between the 3 curve types (Tw Edwards, Montgomery, Jacobian).

**Comment by Bounce23** [comment]
We are still able to use a basepoint encoding for our curve, to fit with the ristretto scalar field. This is the one shown here: 
```
sage: def findBasepoint(prime, A):
....:        F = GF(prime)
....:        E = EllipticCurve(F, [0, A, 0, 1, 0])
....: 
....:        for uInt in range(1, 1e3):
....:          u = F(uInt)
....:          v2 = u^3 + A*u^2 + u
....:          if not v2.is_square():
....:            continue
....:          v = v2.sqrt()
....:          point = E(u, v)
....:          pointOrder = point.order()
....:          if pointOrder > 8 and pointOrder.is_prime():
....:             Q=u^3 + A*u^2 + u
....:             return u, Q, sqrt(Q), point
....:
sage: res = findBasepoint(x, 346598)
sage: res
(17,
 100171752,
 1014685013428497365422144808165958100622560545891891747637198454693655077041,
 (17 : 1014685013428497365422144808165958100622560545891891747637198454693655077041 : 1))
```

The difference is that we make use of an Edwards Y, without the need to specify the 'unique' basepoint stemming from (u-1)/(u+1), as this is for thee 25519 Montgomery fast scalar multiplication. Whereas ours is Edwards points encoded as field elements. 
The chosen Y value for a base point passes all the tests.

**Comment by Bounce23** [comment]
Furthering the above comment: the base point for twisted Edwards, Y = 100171752, is not chosen arbitrarily; and is in line with the safe curves criteria as shown [here](safecurves.cr.yp.to/rigid.html), which is found in the code at line 422 [here](https://github.com/dusk-network/dusk-zerocaf/blob/master/sage_codes/Safe%20Curve%20code). This is set such that the curve maintains rigidity and to allow y(P) as a ladder coordinate.

**Comment by CPerezz** [comment]
This is closed since we take part of this on #82
